### PR TITLE
C# 7 and string interpolation with typeof in SQL queries

### PR DESCRIPTION
**Pull request by warpech**
Implements the advice from #29 and #25.

@Starcounter/sc-apps please review.

### CODE REVIEW

File: coding_style/csharp_code_formatting.md [hunk 1]
```diff
@@ -1,8 +1,33 @@
...
+It is encouraged to use C# 7.0 in Starcounter apps. Among the new features, [expression bodied accessors](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7#more-expression-bodied-members) are very handy in view-model code-behind.
+
+Apps that make use of C# 7.0 should include the following note in `README.md`:
+
+```
+## Requirements
+
+This repository uses features from C# 7. Thus, **Visual Studio 2017** is required to build.
+
+If you really want to use VS 2015 and you don't care about IntelliSense warnings, you can build by installing [Microsoft.Net.Compilers](https://www.nuget.org/packages/Microsoft.Net.Compilers/) in the project using NuGet (`Install-Package Microsoft.Net.Compilers`). That will make VS use the C# compiler provided in the package instead of the default one.
```
Comment (Mackiovello) [inline coding_style/csharp_code_formatting.md]: I would rephrase this one.

To build the repository with  Visual Studio 2015, install [Microsoft.Net.Compilers](https://www.nuget.org/packages/Microsoft.Net.Compilers/) in the project using NuGet (`Install-Package Microsoft.Net.Compilers`). That will make VS use the C# compiler provided in the package instead of the default one. IntelliSense warnings will still be displayed for the C# 7 syntax.

Comment (joozek78) [inline coding_style/csharp_code_formatting.md]: Agree, but 
> IntelliSense warnings will still be displayed for the C# ~~7~~ 6 syntax.
I would highlight this information. It's on the very end and if someone missed it he could get really confused by the error messages

File: starcounter-api-usage.md [hunk 2]
```diff
@@ -16,4 +16,36 @@ Incorrect:
...
+WHERE s.{nameof(SomebodyGroup.Name)} LIKE ?", obj)
+```
+
+It is refactorable and in general it provides better information for coding tools as compared to the table names given as strings.
+
+Warning: if the table name includes keywords such as "order", the above code will not produce a correct query. You will have to escape the keyword in double quotes in one of the following ways:
+
+```c#
+Db.SQL("SELECT * FROM Customer.\"Order\""); //fast but not refactorable
+Db.SQL($"SELECT * FROM \"{typeof(Customer).FullName.Replace(".", "\".\"")}\""); //slow but refactorable
```
Comment (joozek78) [inline starcounter-api-usage.md]: wait, can't we just use `$"SELECT * FROM \"{typeof(Customer).FullName}\""`? I thought that quoting the entire name is sufficient

File: starcounter-api-usage.md [hunk 3]
```diff
@@ -16,4 +16,36 @@ Incorrect:
...
+```
+
+Keep in mind that `typeof` is resolved in the run time. For performance critical code, use `nameof`, which is resolved in the compile time:
```
Comment (joozek78) [inline starcounter-api-usage.md]: I fear that people will use `nameof` instead of using prepared statements. Of course, we don't have them (yet?), but if construction of string takes too much time, then simply preparing it beforehand and reusing it later should solve the performance and make the the string refactorable.

So maybe:
> Keep in mind that `typeof` is resolved in the run time. If a simple query is executed very often (put some rule of thumb number here?), then construction of the SQL string itself could be a performance problem. To avoid it, simply construct the query once, and then reuse it.
```c#
var query = $"SELECT s FROM {typeof(SomebodyGroup).FullName} s WHERE s.{nameof(SomebodyGroup.Name)} LIKE ?";
var names = new[] {"a", "very", "long", "list"};
foreach(var name in names)
{
  var group = Db.SQL<SomebodyGroup>(query, name);
  DoSomethingAboutIt(group);
}
```

NOTE: The nameof pattern introduces several problems, one them is that if you move your DB classes to a different namespace, the query can be invalid (it will refer to old namespace), while the code still compiles

File: starcounter-api-usage.md [hunk 4]
```diff
@@ -16,4 +16,36 @@ Incorrect:
...
+
+```c#
+Db.SQL<SomebodyGroup>($"SELECT s FROM {nameof(Simplified)}.{nameof(Ring2)}.{nameof(SomebodyGroup)} s
+WHERE s.{nameof(SomebodyGroup.Name)} LIKE ?", obj)
+```
+
+If the table is in the same namespace as the current app, then you can use the short table name:
```
Comment (joozek78) [inline starcounter-api-usage.md]: what is the namespace of the app? Did you mean the current class (one holding the method calling the `Db.SQL`)? Using this shortened form can cause trouble (imagine someone moved this piece of code to a different class), I would inform the reader that it's possible, but advised strongly against it.

File: starcounter-api-usage.md (hunk 2 above)
Comment (warpech) [inline starcounter-api-usage.md]: Not in 2.3.x. See: https://github.com/Starcounter/Guidelines/issues/25#issuecomment-301049393

File: coding_style/csharp_code_formatting.md (hunk 1 above)
Comment (warpech) [inline coding_style/csharp_code_formatting.md]: @joozek78 can you explain? I think @Mackiovello means that the usage of C# **7** in VS 2015 will result in warnings. There is no problem with C# 6.

Comment (joozek78) [inline coding_style/csharp_code_formatting.md]: IntelliSense will still evaluate the source as C# 6 and display relevant errors. 

> IntelliSense warnings will still be displayed for the C# 7 syntax.
>  👍 1

I understand this sentence like this - intellisense will still display errors as if it were C# 7 syntax.
Maybe rephrase to "IntelliSense will interpret the source as C# 6. C# 7 features will be highlighted as errors"

File: starcounter-api-usage.md (hunk 3 above)
Comment (warpech) [inline starcounter-api-usage.md]: Great advice, thanks! I will change it.

File: starcounter-api-usage.md (hunk 4 above)
Comment (warpech) [inline starcounter-api-usage.md]: > what is the namespace of the app? Did you mean the current class (one holding the method calling the Db.SQL)?

It is an official Starcounter feature, there is no point in having a feature and strongly advising against it.

The discussion about the future of short names is here: https://github.com/Starcounter/RebelsLounge/issues/111. Please contribute there.

Comment (joozek78) [inline starcounter-api-usage.md]: Then I'm advising against carefree usage of this feature. It can introduce bugs which are hard to spot

Comment (joozek78) [inline starcounter-api-usage.md]: BTW my question stands - what is the namespace of the current app? I thought the short form uses namespace of the calling class

Comment (warpech) [inline starcounter-api-usage.md]: It is the current class (one holding the method calling the `Db.SQL`).

Comment (warpech) [inline starcounter-api-usage.md]: I will remove the guideline. If it is not in the docs, why bother recommending it in our guidelines.

**Comment by warpech** [comment]
Thank you so much for the reviews. I consider it good for merging now.
